airbyte-server/src/main/java/io/airbyte/server/ServerApp.java

```diff
@@ -256,6 +266,85 @@ public static ServerRunnable getServer(final ServerFactory apiFactory, final Con
     }
   }
 
+  /**
+   * Check that each spec in the database has a spec. If it doesn't, add it. If it can't be added,
+   * track the failure in Segment. The goal is to try to end up in a state where all definitions in
+   * the db contain specs, and to understand what is stopping us from getting there.
+   *
+   * @param configRepository - access to the db
+   * @param schedulerClient - scheduler client so that specs can be fetched as needed
+   */
+  private static void migrateAllDefinitionsToContainSpec(final ConfigRepository configRepository,
+                                                         final SynchronousSchedulerClient schedulerClient)
+      throws Exception {
+    for (final StandardSourceDefinition sourceDef : configRepository.listStandardSourceDefinitions()) {
+      try {
+        if (sourceDef.getSpec() == null) {
+          LOGGER.info(
+              "migrateAllDefinitionsToContainSpec - Source Definition {} does not have a spec. Attempting to retrieve spec...",
+              sourceDef.getName());
+          final SynchronousResponse<ConnectorSpecification> getSpecJob = schedulerClient
+              .createGetSpecJob(sourceDef.getDockerRepository() + ":" + sourceDef.getDockerImageTag());
+          if (getSpecJob.isSuccess()) {
+            LOGGER.info(
+                "migrateAllDefinitionsToContainSpec - Spec for Source Definition {} was successfully retrieved. Writing to the db...",
+                sourceDef.getName());
+            final StandardSourceDefinition updatedDef = Jsons.clone(sourceDef).withSpec(getSpecJob.getOutput());
+            configRepository.writeStandardSourceDefinition(updatedDef);
+            LOGGER.info(
+                "migrateAllDefinitionsToContainSpec - Spec for Source Definition {} was successfully written to the db record.",
+                sourceDef.getName());
+          } else {
+            LOGGER.info(
+                "migrateAllDefinitionsToContainSpec - Failed to retrieve spec for Source Definition {}.",
+                sourceDef.getName());
+            throw new RuntimeException(String.format("Failed to retrieve spec for Source Definition %s", sourceDef.getName()));
```

Comment:
Unfortunately if we hit this case where the getSpecJob is unsuccessful, we won't have much useful information to report to Segment besides the fact that it failed for that docker image. Is there any way to glean more info about how the temporal job failed in this case? The only thing I could find was getting the log path which doesn't sound very useful to have in segment.

Reply:
i think the `Synchronous` response object that is `getSpecJob` can help. `JobConverters.getLogRead(getSpecJob.getMetadata().getLogPath(getLogPath));`. essentially that object has a log path. and then the converters class has code to read from that log path. it might be too much info to send though... you'd have to play with it a bit to see if it is useful.